### PR TITLE
Add tests for multiple routes

### PR DESCRIPTION
**Pull request by yschimke**
Test for IPv4, IPv6 multiple routes.  When the RouteDatabase is updated, how many follow up requests are made.

### CODE REVIEW

**Comment by yschimke** [comment]
Using this to test 

https://github.com/square/okhttp/issues/7553
and
https://github.com/square/okhttp/issues/7530

File: okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt [hunk 1]
```diff
@@ -0,0 +1,240 @@
+package okhttp3
```
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:1]: supernit: copyright header

File: okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+import okhttp3.internal.http2.ErrorCode
+import okhttp3.testing.Flaky
+import okhttp3.testing.PlatformRule
+import okhttp3.tls.internal.TlsUtil.localhost
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.RegisterExtension
+
+@Flaky
```
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:22]: eep

File: okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+  private lateinit var server1: MockWebServer
+  private lateinit var server2: MockWebServer
+
+  private var listener = RecordingEventListener()
+
+  private val handshakeCertificates = localhost()
+
+  val dns = FakeDns()
+
+  val ipv4 = InetAddress.getByName("192.168.1.1")
```
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:42]: This is suspect!

Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:42]: Ahhh I see what’s happening. Works for me.

Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:42]: Wanna use one that’s obviously a lie, like `255.255.255.255` ?

Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:42]: (this one looks like a local testing server)

File: okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+      .apply {
+        retryOnConnectionFailure = true
+      }
+      .build()
+
+    executeSynchronously(request)
+      .assertBody("body")
+
+    assertThat(client.routeDatabase.failedRoutes).isEmpty()
+    // TODO check if we expect a second request to server1, before attempting server2
```
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:120]: Yep

Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:120]: (I think that’s the behavior we should test for, though we could create an issue and change that)

File: okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:22]: Is this due to raciness in fast fallback? I wonder if we can add something to make it less racy, like changing the timeouts to MAX_VALUE or 0 or something

File: okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+      val response = call.execute()
+      val bodyString = response.body.string()
+      RecordedResponse(request, response, null, bodyString, null)
+    } catch (e: IOException) {
+      RecordedResponse(request, null, null, null, e)
+    }
+  }
+}
+
+class SpecificHostSocketFactory(
```
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:218]: Promote to its own file?

Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:218]: This is a great idea BTW

**Review by swankjesse — APPROVED** [review]

File: okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt (hunk 2 above)
Comment (yschimke) [reply okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:22]: Sorry, not Flaky. I'm just using it for logging the tracing.

**Pull request merged**